File: backend/execution/ProjectManager.js
class ProjectManager {
    constructor(s3, dynamoDB, docker, tableName, s3BucketName, homeDir) {
        this.s3 = s3;
        this.dynamoDB = dynamoDB;
        this.docker = docker;
        this.tableName = tableName;
        this.s3BucketName = s3BucketName;
        this.homeDir = homeDir;
    }

    async getProjectFiles(userId, projectName) {
        const projectMetadata = await this.getProjectMetadata(userId, projectName);

        if (!projectMetadata) {
            return {};
        }

        return await this.buildProject(projectMetadata);
    }

    async getProjectMetadata(userId, projectName) {
        const params = {
            TableName: this.tableName,
            KeyConditionExpression: 'PK = :pk AND begins_with(SK, :sk_prefix)',
            ExpressionAttributeValues: {
                ':pk': `user#${userId}`,
                ':sk_prefix': `project#${projectName}_file#`
            }
        };

        try {
            const response = await this.dynamoDB.query(params).promise();
            return response.Items || [];
        } catch (error) {
            console.error(error);
            return [];
        }
    }

    async buildProject(dynamoData) {
        if (!dynamoData) {
            console.error('Invalid files data');
            return {};
        }

        const fileMetadata = {};

        for (const file of dynamoData) {
            const filePath = file.SK.split('#')[2];

            fileMetadata[filePath] = {
                fileLink: file.fileLink,
                language: file.language,
                Key: file.s3_key
            };

            try {
                const s3ResponseObject = await this.s3.getObject({ Bucket: this.s3BucketName, Key: file.s3_key }).promise();
                fileMetadata[filePath].content = s3ResponseObject.Body.toString('utf-8');
            } catch (error) {
                console.error(error);
            }
        }

        return fileMetadata;
    }

    async saveFile(userId, projectName, containerId, filePath, fileContent = '', language = 'python') {
        const s3_key = `${userId}/${projectName}/${filePath}`;
        let s3Response;

        try {
            s3Response = await this.s3.putObject({
                Bucket: this.s3BucketName,
                Key: s3_key,
                Body: fileContent
            }).promise();
        } catch (error) {
            throw new Error(`Error uploading to S3: ${error}`);
        }

        const item = {
            PK: `user#${userId}`,
            SK: `project#${projectName}_file#${filePath}`,
            fileLink: s3Response.Location || '',
            Key: s3_key,
            language: language,
            containerId: containerId
        };

        try {
            const dynamoResponse = await this.dynamoDB.put({ TableName: this.tableName, Item: item }).promise();
            return { s3Upload: s3Response, dynamoResponse: dynamoResponse };
        } catch (error) {
            throw new Error(`Error updating DynamoDB: ${error}`);
        }
    }
}

module.exports = ProjectManager;

File: backend/flask-execution/ProjectManager.py
import os
import tarfile
import uuid
from typing import List


class ProjectManager:
    def __init__(self, s3_client, dynamo_db, docker_client, table_name, s3_bucket_name, home_dir):
        self.s3_client = s3_client
        self.dynamo_db = dynamo_db
        self.docker_client = docker_client
        self.table_name = table_name
        self.s3_bucket_name = s3_bucket_name
        self.home_dir = home_dir


    def get_project_files(self, user_id: str, project_name: str) -> dict:
        proj_metadata = self.get_project_metadata(user_id, project_name)
        if not proj_metadata:
            return {}

        files = self.build_project(proj_metadata)
        return files


    def get_project_metadata(self, user_id: str, project_name: str) -> List[dict]:
        response = self.dynamo_db.query(
            TableName=self.table_name,
            KeyConditionExpression='PK = :pk AND begins_with(SK, :sk_prefix)',
            ExpressionAttributeValues={
                ':pk': {'S': f'user#{user_id}'},
                ':sk_prefix': {'S': f'project#{project_name}_file#'},
            },
        )
        if 'Items' in response:
            return response['Items']
        else:
            return []


    def build_project(self, dynamo_data: dict) -> dict:
        if not dynamo_data:
            print('Invalid files data:')
            return {}

        fileMetaData = {}

        for file in dynamo_data:
            filePath = file['SK']["S"].split('#')[2]

            fileMetaData[filePath] = {'fileLink': file["fileLink"]["S"],
                                         'language': file["language"]["S"],
                                         's3_key': file["s3_key"]["S"]}

            s3_response_object = self.s3_client.get_object(Bucket=self.s3_bucket_name, Key=file['s3_key']["S"])
            object_content = s3_response_object['Body'].read()

            object_content = object_content.decode('utf-8')

            fileMetaData[filePath]['content'] = object_content

        return fileMetaData


    def save_file(self, user_id: str, project_name: str, container_id: str, file_path: str, file_content='',
                  language='python') -> dict:
        """
        upload a single file to s3 and store metadata in dynamodb
        """
        s3_key = f'{user_id}/{project_name}/{file_path}'
        try:
            s3_response = self.s3_client.put_object(Bucket=self.s3_bucket_name, Key=s3_key, Body=file_content)
        except Exception as e:
            raise(f'Error uploading to S3: {e}')

        # Store the metadata in DynamoDB
        item = {
            'PK': {'S': f'user#{user_id}'},
            'SK': {'S': f'project#{project_name}_file#{file_path}'},
            'fileLink': {'S': s3_response.get('location', '')},
            's3_key': {'S': s3_key},
            'language': {'S': language},
            'container_id': {'S': container_id},
        }

        try:
            dynamo_response = self.dynamo_db.put_item(
                TableName=self.table_name,
                Item=item
            )
        except Exception as e:
             raise(f'Error updating DynamoDB: {e}')

        return {'s3Upload': s3_response, 'dynamoResponse': dynamo_response}